affinity/cpp-20/d0796r3.md

```diff
@@ -589,17 +630,19 @@ The `affinity_query` class template provides an abstraction for a relative affin
 
 ## Free functions
 
-### `this_system::get_resources`
+### `this_system::discover_topology`
+
+The free function `this_system::discover_topology` is provided for discovering the **system topology**.
 
-The free function `this_system::get_resources` is provided for retrieving the `execution_resource`s which encapsulate the hardware platforms available within the system. We refer to these resources as the *system level resources*.
+	const execution_resource discover_topology();
 
-	std::vector<execution_resource> resources() noexcept;
+*Returns:* An `execution_resource` object exposing the **system execution resource**.
 
-*Returns:* An `std::vector` containing all *system level resources*.
+*Requires:* If `this_system::discover_topology().size() > 0`, `this_system::discover_topology()[0]` be the `execution_resource` use by `std::thread`. Calls to `this_system::discover_topology()` may not introduce a data race with any other call to `this_system::discover_topology()`.
```

Comment:
Does that mean the implementation is responsible to implementing mutex or similar to ensure no data races?
This means the call to discover topology is (even more) expensive. Seems to me, since this is a function that will be called by the library programmer or the programmer on controlled situations, its better for them to handle the control of accessing this function.

Reply:
Yes, the current wording requires that `discover_topology` remain race free with respect to other calls to `discover_topology`. This prevents two simultaneous calls to `discover_topology` from interfering with each other and potentially trying to initialise or finalise the same third-party APIs at the same time. We felt this a fairly low burden to users as topology discovery will generally happen away from the critical path and this will only incur a cost if users were to have multiple threads potentially capable of performing topology discovery. And it means users can safely do this if they want.

This doesn't cover the case where users may manually call a third-party API which invokes the same APIs as `discover_topology`, however, we felt this is not something we can protect users from, and implementations should accurately document any dependencies they have.

This also doesn't currently cover is any potential races between using a resource being used (for example in some execution context) whilst performing topology discovery. This is a slightly trickier case to resolve which we will have to look at.